timemachine/potentials/nonbonded.py

```diff
@@ -155,6 +159,56 @@ def nonbonded_block(xi, xj, box, params_i, params_j, beta, cutoff):
     return jnp.sum(nonbonded_block_unsummed(xi, xj, box, params_i, params_j, beta, cutoff))
 
 
+import itertools
+
+
+def nonbonded_block_rescaled(xi, xj, box, params_i, params_j, beta, cutoff, anchors_i, groups_i):
+    """
+
+    Soft-core variation where w_i is applied by means of rescaling dummy atoms relative to an
+    arbitrary choice of the anchoring atom. If w_i is 0, then atom i is fully interacting.
+    If w_i is 1, then atom_i is fully non-interacting. If 0 < w_i < 1 then atom_i is scaled
+    relative to the location of the anchoring atom for the group that atom_i belongs to. Atoms
+    that are not part of the groups_i ignore their w_i values.
+
+    Note that we assume that charges have *already* been balanced correctly in the upstream
+    interpolation code.
+
+    Example:
+
+    anchors_i = [0, 5, 9]
+    groups_i = [[3,4], [7], [6,8,2]]
+    w_i = [[0.2, 0.2], [0.8], [0.2, 0.2]]
+
+    """
+
+    xi = jnp.array(xi)
+    params_i = jnp.array(params_i)
+
+    assert len(anchors_i) == len(groups_i)
+
+    # transform xi
+    xi_a = xi[np.repeat(anchors_i, [len(x) for x in groups_i])]
+
+    group_idxs = jnp.array(list(itertools.chain(*groups_i)))
+
+    excluded_idxs = list(set(range(len(xi))).difference(group_idxs.tolist()))
```

Comment:
nit: can be optimized to avoid Python loops (OK to leave for later)

Reply:
stale